Btrix Operator Crawl Management

ikreymer opened the pull request:
Next phase of the operator refactor (#550), the BtrixJob metacontroller will manage crawl and profile objects directly.
This works involves the following:
- [x] Updating BtrixJob crd to manage statefulsets
- [x] Updating status spec to reflect job status
- [x] Ensuring crawl stats (from crawl_job.py) are properly updated by operator
- [x] Ensuring deleting BtrixJob deletes the crawl sts + services
- [x] Ensuring BtrixJob is cleaned up automatically after finish(use `ttlSecondsAfterFinished` similar to k8s Job?)
- [x] Ensure updating scale dynamically works
- [x] Split operator into separate container, ensure current tests pass
- [x] Ensuring graceful stop, forced rollover, and cancel job work
- [x] Figure out support for Profiles, (ProfileJob vs CrawlJob)
- [x] Figure out support for schedule job (CronJob -> Job -> BtrixJob for now)

Cleanup:
- [x] Remove *_job.py files
- [x] Ensure operator class uses existing crawls / crawlconfig wrappers
- [x] Remove separate .k8s package (only supporting k8s)

ikreymer committed:
switch operator to create sts for crawler / redis directly, remove job

ikreymer committed:
add scale

ikreymer committed:
work

ikreymer committed:
fix typo

ikreymer committed:
crawls working with btrixjob
- update status to reflect current crawl state
- mark crawl as running, failed, complete when finished
- store finished status when crawl is complete

ikreymer committed:
add auto-delete of BtrixJob via 'ttlSecondsAfterFinished'
also delete pvcs until autodelete supported via statefulset (k8s >1.27)
ensure filesAdded always set correctly, keep counter in redis, add to status display

ikreymer committed:
fix scale updates

ikreymer committed:
separate container / fix tests:
- run operator api in separate container but in same pod, use same as ima
- readd update_and_prepare_db (run for one worker only)
- add healthcheck
- fix tests?

ikreymer committed:
lint: readd disable dupe code check

ikreymer committed:
btrix job mutation:
- support updating scale, forcing rollover, stop via patching btrix-job
- support cancel via deletion
- requires hack to content-length for patching

ikreymer committed:
rename btrixjob -> crawljob
add profilejob and profilecontroller

ikreymer committed:
- profilejobs working, ensure proper cleanup
- ensure content-type setting works correctly
- add base k8s functionality to k8sapi

ikreymer committed:
refactor/cleanup:
- remove k8s package
- merge k8sman and basecrawlmanager into crawlmanager
- move templates, k8sapi, utils into root package
- delete all *_job.py files
- rename btrix_job.yaml -> crawl_job.yaml
- remove dt_now, ts_now from crawls, now in utils

ikreymer committed:
cleanup crawlmanager
cleanup operator, fail if configmap not found

ikreymer committed:
support for schedule jobs!
- add crawl_cron_job for launching CronJob, which starts via main_scheduled_job entrypoint to create a CrawlJob
- add main_scheduled_job entrypoint for scheduled jobs
- all db operations happen in crawl/crawlconfig/org files
- refactor: move shared crawl/crawlconfig/org functions that use the db to be importable directly,
including get_crawl_config, add_new_crawl, inc_crawl_stats
- operator: use service hook endpoint, set port in values.yaml

ikreymer committed:
role binding: more secure setup, don't allow crawler namespace any k8s permissions
- move cronjobs to be created in default namespace
- grant default namespace access to create cronjobs in default namespace
- remove role binding from crawler namespace

ikreymer committed:
scale: ensure changing scale of crawl also updates it in the db

ikreymer committed:
crawlconfig map check: move check to separate run method, otherwise configmap created before check on new crawl

leepro commented:
Overall, Refactoring went well. Looks neat and more simpler. Nice job!
Should rename files having `btirxjobs_` with `crawljobs_`?

ikreymer committed:
additional tweaks to templates/operator:
- templates: split crawler and redis sts into seperate yaml file (in case need to load one or other separately)
- operator: optimization: attempt to reduce automerging, by reusing volumeClaimTemplates from existing children, as these may have additional props added
- templates: remove unsupported features for now, remove properties from status/spec to reduce chance of unnecessary merges
- tempaltes: fix typo in store filename variable!
- crawlmanager: remove unused code

ikreymer committed:
stats / redis optimization:
- don't update stats in mongodb on every operator sync, only when crawl is finished
- for api access, read stats directly from redis to get up-to-date stats
- move get_page_stats() to utils, add get_redis_url() to k8sapi to unify access

tw4l commented:
Very nice! Architecture is a lot cleaner now and the start time speed up for crawls is great! As discussed elsewhere, I'll submit a separate PR for the migration after this PR is merged.

ikreymer commented:
> Overall, Refactoring went well. Looks neat and more simpler. Nice job! Should rename files having `btirxjobs_` with `crawljobs_`?

I think btrix_jobs might be ok as there's now crawljob and profilejob? Sort of keeps it more general
